keps/sig-node/4382-dra-numeric-model-simple-counter/README.md

```diff
@@ -0,0 +1,815 @@
+instance. There are no restrictions on how a resource instance can be split up
+among claims.
+
+### Risks and Mitigations
+
+Adding another numeric model doesn't add any additional risks. There is a
+slight performance impact when checking whether the builtin controller of the
+numeric model needs to do something for a claim or NodeResourceCapacity object,
+but that impact should be minimal when the model is not in use.
+
+## Design Details
+
+Everything related to this numeric model is hosted in
+`k/k/staging/src/k8s.io/dynamic-resource-allocation`:
+
+- `numeric/counter`: the implementation of the builtin controller
+- `apis/counter`: the versioned type(s)
+- `apis/internal/counter`: the internal type for conversion
+
+The only modification elsewhere is registering the builtin controller in the
+dynamic resource scheduler plugin if (and only if) the model is enabled.
+
+### Capacity type
+
+```
+// Capacity defines how much of certain items are available respectively
+// required.
+type Capacity struct {
+	// ObjectMeta may be set when publishing capacity, but is not required.
+	// It is ignored when matching requests against available capacity.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// TypeMeta must have non-empty kind and apiVersion.
+	metav1.TypeMeta `json:",inline"`
+
+	// Count is the total number of available items per node when
+	// publishing capacity and the required number of items when requesting
+	// resources.
+	Count int64 `json:"count"`
+}
+```
+
+### Claim parameter type
+
+There is no separate type because the same `Count` field also makes sense when
+requesting items. The examples in the numeric parameters KEP show how this then
+gets used by a DRA driver.
+
+### Allocation result
+
+The following object gets serialized as JSON and stored in a claim's
+"status.allocation.resourcehandle.data" field:
+
+```
+// AllocationResult captures class and claim parameters and records which node
+// capacity instance was used. The JSON dump of this gets stored in the driver
+// handle field of an allocated claim.
+type AllocationResult struct {
+	// TypeMeta must have non-empty kind and apiVersion.
+	metav1.TypeMeta `json:",inline"`
+
+	// ClassParameters are the parameters from the time that the claim
+	// was allocated. They can be arbitrary setup parameters that are
+	// ignore by the counter controller.
+	ClassParameters runtime.RawExtension `json:"classParameters,omitempty"`
+
+	// ClaimParameters are the parameters from the time that the claim was
+	// allocated. Some of the fields were used by the counter controller to
+	// allocated resources, others can be arbitrary setup parameters.
+	ClaimParameters runtime.RawExtension `json:"claimParameters,omitempty"`
+
+	// NodeName is the name of the node providing the necessary resources.
+	NodeName string `json:"nodeName"`
+
+	// DriverName is the name of the driver on that node.
+	DriverName string `json:"driverName"`
```

Comment:
Can there be multiple drivers on a specific node. Does this handle allocations by multiple drivers?

Reply:
The NodeResourceCapacity is per-node, per-driver, per-resource-instance. This field is necessary for the numeric model to determine which capacity gets freed when this allocation is removed.